src/ekka_mnesia.erl

```diff
@@ -366,26 +366,47 @@ wait_for(tables) ->
 %% Transaction API
 %%--------------------------------------------------------------------
 
--spec ro_transaction(fun(() -> A)) -> t_result(A).
-ro_transaction(Fun) ->
-    %% TODO:
-    %% case ekka_rlog:role() of
-    %%     core      -> ok;
-    %%     replicant -> _ = ekka_rlog_status:upstream() % assert that replica is fully up
-    %% end,
-    mnesia:transaction(fun ekka_rlog_activity:ro_transaction/1, [Fun]).
+-spec ro_transaction(ekka_rlog:shard(), fun(() -> A)) -> t_result(A).
+ro_transaction(Shard, Fun) ->
+    case ekka_rlog:role() of
+        core ->
+            mnesia:transaction(fun ekka_rlog_activity:ro_transaction/1, [Fun]);
+        replicant ->
+            ?tp(ekka_ro_transaction, #{role => replicant}),
+            case ekka_rlog_status:upstream(Shard) of
+                {ok, AgentPid} ->
+                    Ret = mnesia:transaction(fun ekka_rlog_activity:ro_transaction/1, [Fun]),
+                    %% Now we check that the agent pid is still the
+                    %% same, meaning the replicant node haven't gone
+                    %% through bootstrapping process while running the
+                    %% transaction and it didn't have a chance to
+                    %% observe the stale writes.
+                    case ekka_rlog_status:upstream(Shard) of
+                        {ok, AgentPid} ->
+                            Ret;
+                        _ ->
+                            %% Restart transaction. If the shard is
+                            %% still disconnected, it will become an
+                            %% RPC call to a core node:
+                            ro_transaction(Shard, Fun)
+                    end;
+                disconnected ->
+                    ro_trans_rpc(Shard, Fun)
+            end
+    end.
 
--spec transaction(fun((...) -> A), list()) -> t_result(A).
-transaction(Fun, Args) ->
-    ekka_rlog:call_backend_rw_trans(transaction, [Fun, Args]).
+-spec transaction(ekka_rlog:shard(), fun((...) -> A), list()) -> t_result(A).
+transaction(Shard, Fun, Args) ->
```

Comment:
The main reason to make this API change is the plugins: they add and remove mnesia tables in the runtime. So in theory, a situation where different core nodes have a different set of tables. This way replicas will know in advance which core node they should talk to.